### PR TITLE
Validity Window Syncer

### PR DESCRIPTION
**Pull request by viquezclaudio**
The Validity Window Syncer is a component that operates between the macro sync and the live sync
Its purpose is to sync all the historic transactions that happened during the current validity window
This is because, as a validator, we need to enforce the "tx in validity window" check when producing blocks

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

File: blockchain/src/blockchain/wrappers.rs [hunk 1]
```diff
@@ -258,34 +258,27 @@ impl Blockchain {
         self.state().accounts.tree.get_missing_range(txn)
     }
 
-    /// Check if all blockchain conditions to run a validator are met.
-    /// These are:
-    /// - can verify/extend history tree
-    /// - can enforce transaction validity window
-    /// The second subsumes the first.
-    /// All conditions are checked against the next block (which potentially hasn't been produced yet).
+    /// Check if we have enough state to check for duplicates during the validity window
```
Comment (jsdanielh) [inline blockchain/src/blockchain/wrappers.rs]:
```suggestion
    /// Check if we have enough state to check for duplicate transactions during the validity window
```

File: blockchain/tests/blockchain.rs [hunk 2]
```diff
@@ -14,6 +14,7 @@ use nimiq_test_utils::{
 };
 
 #[test(tokio::test)]
+#[ignore]
```
Comment (jsdanielh) [inline blockchain/tests/blockchain.rs:17]: Add a `TODO`

File: consensus/src/messages/mod.rs [hunk 3]
```diff
@@ -19,7 +19,7 @@ use thiserror::Error;
 
 use crate::error::SubscribeToAddressesError;
 
-mod handlers;
+pub mod handlers;
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs]: `pub(crate)` maybe?

File: consensus/src/messages/mod.rs [hunk 4]
```diff
@@ -49,6 +49,8 @@ pub const MAX_REQUEST_TRANSACTIONS_BY_ADDRESS: u32 = 1000;
 pub const MAX_REQUEST_TRIE_PROOF: u32 = 1000;
 /// The max number of Block proof requests per peer.
 pub const MAX_REQUEST_BLOCKS_PROOF: u32 = 1000;
+/// The max number of ValidityWindow requests per peer.
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs]:
```suggestion
/// The max number of validity window start requests per peer.
```

File: consensus/src/messages/mod.rs [hunk 5]
```diff
@@ -500,3 +502,29 @@ impl Topic for AddressSubscriptionTopic {
     const NAME: &'static str = "AddressNotification";
     const VALIDATE: bool = false;
 }
+
+/// This request is used to initialize the Validity Window sync process
+/// First we send our current macro head information
+/// The server will provide to us the validity window start proof
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestValidityWindowStart {
+    pub macro_head_number: u32,
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs]: Could you add some `rustdoc`?

File: consensus/src/messages/mod.rs [hunk 6]
```diff
@@ -500,3 +502,29 @@ impl Topic for AddressSubscriptionTopic {
...
+    pub macro_head_hash: Blake2bHash,
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs]: Could you add some `rustdoc`?

File: consensus/src/messages/mod.rs [hunk 7]
```diff
@@ -49,6 +49,8 @@ pub const MAX_REQUEST_TRANSACTIONS_BY_ADDRESS: u32 = 1000;
 pub const MAX_REQUEST_TRIE_PROOF: u32 = 1000;
 /// The max number of Block proof requests per peer.
 pub const MAX_REQUEST_BLOCKS_PROOF: u32 = 1000;
+/// The max number of ValidityWindow requests per peer.
+pub const MAX_REQUEST_VALIDITY_WINDOW_START: u32 = 1000;
```
Comment (jsdanielh) [inline consensus/src/messages/mod.rs]: Do we still need it?

File: consensus/src/sync/light/validity_window.rs [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+use super::LightMacroSync;
+#[cfg(feature = "full")]
+use crate::messages::{HistoryChunk, HistoryChunkError, RequestHistoryChunk};
+use crate::sync::{light::sync::ValidityChunkRequest, syncer::MacroSyncReturn};
+
+/// The validity start is defined as the current head - validity_window_blocks
+/// However, if the validity start is located in the current epoch
+/// we will move it to the beggining of the current epoch,
+/// in order to sync the full epoch and being able to verify the history root.
+pub fn compute_validity_start(macro_head: u32) -> u32 {
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs]: This function logic I thin could be simpler

File: consensus/src/sync/light/validity_window.rs [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        self.start_validity_chunk_request(
+            peer_id,
+            verifier_block_number,
+            expected_root,
+            validity_window_bn,
+            election_in_window,
+        );
+    }
+
+    pub fn poll_validity_window_chunks(
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs:168]: Could you add some `rustdoc`?

File: consensus/src/sync/light/validity_window.rs [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<MacroSyncReturn<TNetwork::PeerId>>> {
+        while let Poll::Ready(Some(Ok((request, result, peer_id)))) =
+            self.validity_queue.poll_next_unpin(cx)
+        {
+            log::trace!(peer=%peer_id, chunk_index= request.chunk_index, block_number =request.block_number,  "Processing response from validity queue");
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs]:
```suggestion
            log::trace!(peer=%peer_id, chunk_index=request.chunk_index, block_number=request.block_number,  "Processing response from validity queue");
```

File: consensus/src/sync/light/validity_window.rs [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+                        // Get ready for requesting the next chunk
+                        let mut chunk_index = peer_request.chunk_index + 1;
+
+                        if epoch_complete {
+                            // We need to check if there was an election in between, if so, we need to proceed to the next epoch
+                            if peer_request.election_in_window {
+                                log::trace!(
+                                    "Moving to the next epoch to continue syncing, current epoch {} ",
+                                    Policy::epoch_at(verifier_block_number)
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs]:
```suggestion
                                    current_epoch = Policy::epoch_at(verifier_block_number),
                                    "Moving to the next epoch to continue syncing",
                                    
```

File: consensus/src/sync/light/validity_window.rs [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+
+                        log::trace!(
+                            verifier_bn = verifier_block_number,
+                            chunk_index = chunk_index,
+                            "Adding a new validity window chunk request"
+                        );
+
+                        self.validity_queue.add_ids(vec![(request, None)]);
+                    } else {
+                        // If the chunk doesnt verify we disconnect from the peer
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs]:
```suggestion
                        // If the chunk doesn't verify we disconnect from the peer
```

File: consensus/src/sync/light/validity_window.rs [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+                        log::error!(peer=?peer_id,
+                                    chunk=request.chunk_index,
+                                    verifier_block=request.block_number,
+                                    epoch=request.epoch_number,
+                                    "The validity history chunk didn't verify, disconnecting from peer");
+
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs:278]: I would even ban the peer

File: consensus/src/sync/light/validity_window.rs [hunk 14]
```diff
@@ -0,0 +1,304 @@
+use std::{
+    cmp,
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs]:
```suggestion
```

File: consensus/src/sync/light/validity_window.rs [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+        let macro_head = self.blockchain.read().macro_head();
+
+        let validity_start = macro_head
+            .block_number()
+            .saturating_sub(Policy::transaction_validity_window_blocks());
+
+        let validity_window_bn = if validity_start <= Policy::genesis_block_number() {
+            Policy::genesis_block_number()
+        } else {
+            Policy::election_block_before(validity_start)
```
Comment (jsdanielh) [inline consensus/src/sync/light/validity_window.rs:123]: What if `validity_start` is already an election macro block?

Comment (viquezclaudio) [reply consensus/src/sync/light/validity_window.rs:123]: we need to sync the previous epoch in order to verify the history

File: consensus/src/sync/light/validity_window.rs (hunk 14 above)
Comment (viquezclaudio) [reply consensus/src/sync/light/validity_window.rs]: done

**Review by jsdanielh — APPROVED** [review]

**Comment by jsdanielh** [comment]
@viquezclaudio Could you add that this closes #1473?

**Pull request merged**